cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -348,9 +391,17 @@ private void recordProgramSuspendResume(ProgramId programId, String pid, String
       record = get(key, RunRecordMeta.class);
     }
 
+    // We can also suspend a workflow that is in the starting state
```

Comment:
Since we introduced a starting state, it is possible for a program to be in the starting state and transition to other states besides RUNNING. There are several cases that I've had to handle now:

1. STARTING->SUSPENDED
2. STARTED->FAILED, STARTING->KILLED
3. resuming->RUNNING

I've tried to make these as efficient as possible by only querying the table if the original case did not work.